File: packages/go/http.go
package webpipe

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"
)

// retryableStatuses are retried with exponential backoff.
var retryableStatuses = map[int]bool{429: true, 500: true, 502: true, 503: true, 504: true}

// httpClient is the internal transport: auth headers, retries, error mapping.
type httpClient struct {
	baseURL    string
	apiKey     string
	maxRetries int
	userAgent  string
	client     *http.Client
}

func (h *httpClient) newRequest(ctx context.Context, method, path string, body any) (*http.Request, error) {
	var reader io.Reader
	if body != nil {
		raw, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("webpipe: encode request: %w", err)
		}
		reader = bytes.NewReader(raw)
	}
	req, err := http.NewRequestWithContext(ctx, method, h.baseURL+path, reader)
	if err != nil {
		return nil, fmt.Errorf("webpipe: build request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+h.apiKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", h.userAgent)
	return req, nil
}

func backoff(res *http.Response, attempt int) time.Duration {
	if res != nil {
		if ra := res.Header.Get("Retry-After"); ra != "" {
			if seconds, err := strconv.ParseFloat(ra, 64); err == nil && seconds >= 0 {
				return time.Duration(seconds * float64(time.Second))
			}
		}
	}
	return time.Duration(0.5*float64(uint(1)<<attempt)) * time.Second
}

// do executes the request with retries and decodes the JSON body into out.
func (h *httpClient) do(ctx context.Context, method, path string, body, out any) error {
	var lastErr error
	for attempt := 0; ; {
		req, err := h.newRequest(ctx, method, path, body)
		if err != nil {
			return err
		}
		res, err := h.client.Do(req)
		if err != nil {
			lastErr = fmt.Errorf("webpipe: %s %s: %w", method, path, err)
			if attempt < h.maxRetries {
				attempt++
				if !sleepCtx(ctx, backoff(nil, attempt)) {
					return ctx.Err()
				}
				continue
			}
			return lastErr
		}

		raw, readErr := io.ReadAll(res.Body)
		res.Body.Close()
		if readErr != nil {
			return fmt.Errorf("webpipe: read response: %w", readErr)
		}

		if retryableStatuses[res.StatusCode] && attempt < h.maxRetries {
			attempt++
			if !sleepCtx(ctx, backoff(res, attempt)) {
				return ctx.Err()
			}
			continue
		}
		if res.StatusCode >= 400 {
			return apiErrorFrom(res.StatusCode, raw)
		}
		if out == nil || len(raw) == 0 {
			return nil
		}
		if err := json.Unmarshal(raw, out); err != nil {
			return fmt.Errorf("webpipe: decode response: %w", err)
		}
		return nil
	}
}

func apiErrorFrom(status int, raw []byte) *APIError {
	apiErr := &APIError{StatusCode: status, Message: http.StatusText(status)}
	var body struct {
		Error string `json:"error"`
	}
	if err := json.Unmarshal(raw, &body); err == nil && body.Error != "" {
		apiErr.Message = body.Error
		apiErr.Body = json.RawMessage(raw)
	}
	return apiErr
}

// sleepCtx sleeps for d, returning false if ctx is done first.
func sleepCtx(ctx context.Context, d time.Duration) bool {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-t.C:
		return true
	}
}

File: packages/python/src/webpipe/_http.py
"""Thin httpx wrapper: auth headers, retries with backoff, error mapping.

Private module — never import from outside the package.
"""

import asyncio
import time
from typing import Any, Dict, Optional

import httpx

from .errors import (
    APIError,
    AuthenticationError,
    BadRequestError,
    ForbiddenError,
    NotFoundError,
    RateLimitError,
    ServerError,
)

_STATUS_TO_ERROR = {
    400: BadRequestError,
    401: AuthenticationError,
    403: ForbiddenError,
    404: NotFoundError,
    429: RateLimitError,
}

_RETRYABLE_STATUSES = frozenset({429, 500, 502, 503, 504})


def _backoff(response: Optional[httpx.Response], attempt: int) -> float:
    """Seconds to wait before retry ``attempt`` (1-based)."""
    if response is not None:
        retry_after = response.headers.get("Retry-After")
        if retry_after:
            try:
                return max(0.0, float(retry_after))
            except ValueError:
                pass
    return 0.5 * (2**attempt)


def _raise_for_status(response: httpx.Response) -> None:
    if response.status_code < 400:
        return
    message = response.text
    body: Optional[Any] = None
    try:
        body = response.json()
        if isinstance(body, dict) and body.get("error"):
            message = str(body["error"])
    except ValueError:
        pass
    if response.status_code >= 500:
        error_cls = ServerError
    else:
        error_cls = _STATUS_TO_ERROR.get(response.status_code, APIError)
    raise error_cls(message, response.status_code, body)


def _decode(response: httpx.Response) -> Dict[str, Any]:
    if response.status_code == 204 or not response.content:
        return {}
    payload = response.json()
    return payload if isinstance(payload, dict) else {"data": payload}


def _headers(api_key: str, user_agent: str) -> Dict[str, str]:
    return {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
        "User-Agent": user_agent,
    }


class HttpClient:
    """Synchronous HTTP client with retries."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: float,
        max_retries: int,
        user_agent: str,
    ) -> None:
        self._max_retries = max_retries
        self._client = httpx.Client(
            base_url=base_url,
            headers=_headers(api_key, user_agent),
            timeout=timeout,
        )

    def request(
        self, method: str, path: str, json: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        attempt = 0
        while True:
            response: Optional[httpx.Response] = None
            try:
                response = self._client.request(method, path, json=json)
            except httpx.TransportError:
                if attempt >= self._max_retries:
                    raise
                attempt += 1
                time.sleep(_backoff(None, attempt))
                continue
            if response.status_code in _RETRYABLE_STATUSES and attempt < self._max_retries:
                attempt += 1
                time.sleep(_backoff(response, attempt))
                continue
            _raise_for_status(response)
            return _decode(response)

    def get(self, path: str) -> Dict[str, Any]:
        return self.request("GET", path)

    def post(self, path: str, json: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self.request("POST", path, json)

    def delete(self, path: str) -> Dict[str, Any]:
        return self.request("DELETE", path)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "HttpClient":
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()


class AsyncHttpClient:
    """Asynchronous HTTP client with retries."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: float,
        max_retries: int,
        user_agent: str,
    ) -> None:
        self._max_retries = max_retries
        self._client = httpx.AsyncClient(
            base_url=base_url,
            headers=_headers(api_key, user_agent),
            timeout=timeout,
        )

    async def request(
        self, method: str, path: str, json: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        attempt = 0
        while True:
            response: Optional[httpx.Response] = None
            try:
                response = await self._client.request(method, path, json=json)
            except httpx.TransportError:
                if attempt >= self._max_retries:
                    raise
                attempt += 1
                await asyncio.sleep(_backoff(None, attempt))
                continue
            if response.status_code in _RETRYABLE_STATUSES and attempt < self._max_retries:
                attempt += 1
                await asyncio.sleep(_backoff(response, attempt))
                continue
            _raise_for_status(response)
            return _decode(response)

    async def get(self, path: str) -> Dict[str, Any]:
        return await self.request("GET", path)

    async def post(self, path: str, json: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return await self.request("POST", path, json)

    async def delete(self, path: str) -> Dict[str, Any]:
        return await self.request("DELETE", path)

    async def aclose(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> "AsyncHttpClient":
        return self

    async def __aexit__(self, *exc: object) -> None:
        await self.aclose()
